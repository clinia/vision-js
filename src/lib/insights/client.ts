import { SearchResults } from '@clinia/search-helper';
import { uniq, createDocumentationMessageGenerator } from '../utils';
import {
  Records,
  InsightsClient,
  InsightsClientMethod,
  InsightsClientPayload,
  InsightsClientWrapper,
  Renderer,
  RendererOptions,
  Unmounter,
  WidgetFactory,
} from '../../types';

const getSelectedRecords = (
  records: Records,
  selectedIDs: string[]
): Records => {
  return selectedIDs.map(id => {
    const record = records.find(h => h.id === id);
    if (typeof record === 'undefined') {
      throw new Error(
        `Could not find id "${id}" passed to \`clickedIdsAfterSearch\` in the returned records. This is necessary to infer the absolute position and the query id.`
      );
    }
    return record;
  });
};

const getQueryID = (selectedRecords: Records): string => {
  const queryIDs = uniq(selectedRecords.map(record => record.__queryID));
  if (queryIDs.length > 1) {
    throw new Error(
      'Insights currently allows a single `queryID`. The `IDs` provided map to multiple `queryID`s.'
    );
  }
  const queryID = queryIDs[0];
  if (typeof queryID !== 'string') {
    throw new Error(
      `Could not infer \`queryID\`. Ensure Vision \`clickAnalytics: true\` was added with the Configure widget.`
    );
  }
  return queryID;
};

const getPositions = (selectedRecords: Records): number[] =>
  selectedRecords.map(record => record.__position);

export const inferPayload = ({
  method,
  results,
  records,
  ids,
}: {
  method: InsightsClientMethod;
  results: SearchResults;
  records: Records;
  ids: string[];
}): Omit<InsightsClientPayload, 'eventName'> => {
  const { index } = results;
  const selectedRecords = getSelectedRecords(records, ids);
  const queryID = getQueryID(selectedRecords);

  switch (method) {
    case 'clickedRecordIDsAfterSearch': {
      const positions = getPositions(selectedRecords);
      return { index, queryID, ids, positions };
    }

    case 'convertedRecordIDsAfterSearch':
      return { index, queryID, ids };

    default:
      throw new Error(`Unsupported method passed to insights: "${method}".`);
  }
};

const wrapInsightsClient = (
  aa: InsightsClient | null,
  results: SearchResults,
  records: Records
): InsightsClientWrapper => (
  method: InsightsClientMethod,
  payload: Partial<InsightsClientPayload>
) => {
  if (!aa) {
    const withVisionUsage = createDocumentationMessageGenerator({
      name: 'vision',
    });
    throw new Error(
      withVisionUsage(
        'The `insightsClient` option has not been provided to `vision`.'
      )
    );
  }
  if (!Array.isArray(payload.ids)) {
    throw new TypeError('Expected `ids` to be an array.');
  }
  const inferredPayload = inferPayload({
    method,
    results,
    records,
    ids: payload.ids,
  });
  aa(method, { ...inferredPayload, ...payload } as any);
};

type Connector<TWidgetParams> = (
  renderFn: Renderer<any>,
  unmountFn: Unmounter
) => WidgetFactory<TWidgetParams>;

export default function withInsights(
  connector: Connector<any>
): Connector<unknown> {
  const wrapRenderFn = (
    renderFn: Renderer<RendererOptions<unknown>>
  ): Renderer<RendererOptions<unknown>> => (
    renderOptions: RendererOptions,
    isFirstRender: boolean
  ) => {
    const { results, records, visionInstance } = renderOptions;
    if (results && records && visionInstance) {
      const insights = wrapInsightsClient(
        visionInstance.insightsClient,
        results,
        records
      );
      return renderFn({ ...renderOptions, insights }, isFirstRender);
    }
    return renderFn(renderOptions, isFirstRender);
  };

  return (renderFn: Renderer<RendererOptions<unknown>>, unmountFn: Unmounter) =>
    connector(wrapRenderFn(renderFn), unmountFn);
}
