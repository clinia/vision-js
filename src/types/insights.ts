export type InsightsClientMethod =
  | 'clickedRecordIDsAfterSearch'
  | 'convertedRecordIDsAfterSearch';

export type InsightsClientPayload = {
  eventName: string;
  queryID: string;
  index: string;
  ids: string[];
  positions?: number[];
};

export type InsightsClient = (
  method: InsightsClientMethod,
  payload: InsightsClientPayload
) => void;

export type InsightsClientWrapper = (
  method: InsightsClientMethod,
  payload: Partial<InsightsClientPayload>
) => void;
