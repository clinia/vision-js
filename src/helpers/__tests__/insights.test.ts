import insights, {
  writeDataAttributes,
  readDataAttributes,
  hasDataAttributes,
} from '../insights';

const makeDomElement = (html: string): HTMLElement => {
  const div = document.createElement('div');
  div.innerHTML = html;
  return (div.firstElementChild as HTMLElement) || div;
};

describe('insights', () => {
  test('default behaviour', () => {
    expect(
      insights('clickedRecordIDsAfterSearch', {
        ids: ['3'],
        eventName: 'Add to Cart',
      })
    ).toMatchSnapshot();
  });
});

describe('writeDataAttributes', () => {
  it('should output a string containing data-insights-* attributes', () => {
    expect(
      writeDataAttributes({
        method: 'clickedRecordIDsAfterSearch',
        payload: {
          ids: ['3'],
          eventName: 'Add to Cart',
        },
      })
    ).toMatchSnapshot();
  });
  it('should reject undefined payloads', () => {
    expect(() =>
      // @ts-ignore
      writeDataAttributes({
        method: 'clickedRecordIDsAfterSearch',
      })
    ).toThrowErrorMatchingSnapshot();
  });
  it('should reject non object payloads', () => {
    expect(() =>
      writeDataAttributes({
        method: 'clickedRecordIDsAfterSearch',
        // @ts-ignore
        payload: 2,
      })
    ).toThrowErrorMatchingSnapshot();
  });
  it('should reject non JSON serializable payloads', () => {
    const circularObject: any = { a: {} };
    circularObject.a.circle = circularObject;
    expect(() =>
      writeDataAttributes({
        method: 'clickedRecordIDsAfterSearch',
        payload: circularObject,
      })
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('hasDataAttributes', () => {
  it('should return true when there is a data-insights-method attribute', () => {
    const domElement = makeDomElement(
      `<button
        data-insights-method="clickedRecordIDsAfterSearch"
        data-insights-payload='{"ids":["3"],"eventName":"Add to Cart"}'
        > Add to Cart </button>`
    );

    expect(hasDataAttributes(domElement)).toBe(true);
  });
  it("should return false when there isn't a data-insights-method attribute", () => {
    const domElement = makeDomElement(
      `<button
        data-insights-payload='{"ids":["3"],"eventName":"Add to Cart"}'
        > Add to Cart </button>`
    );

    expect(hasDataAttributes(domElement)).toBe(false);
  });
});

describe('readDataAttributes', () => {
  describe('on handwritten data-insights-* attributes', () => {
    let domElement: HTMLElement;

    beforeEach(() => {
      const payload = btoa(
        JSON.stringify({ ids: ['3'], eventName: 'Add to Cart' })
      );
      domElement = makeDomElement(
        `<button
        data-insights-method="clickedRecordIDsAfterSearch"
        data-insights-payload="${payload}"
        > Add to Cart </button>`
      );
    });

    it('should extract the method name', () => {
      expect(readDataAttributes(domElement).method).toEqual(
        'clickedRecordIDsAfterSearch'
      );
    });

    it('should extract the payload and parse it as a json object', () => {
      expect(readDataAttributes(domElement).payload).toEqual({
        ids: ['3'],
        eventName: 'Add to Cart',
      });
    });

    it('should reject invalid payload', () => {
      domElement = makeDomElement(
        `<button
        data-insights-method="clickedRecordIDsAfterSearch"
        data-insights-payload='xxx'
        > Add to Cart </button>`
      );
      expect(() =>
        readDataAttributes(domElement)
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('on data-insights-* attributes generated with insights helper', () => {
    let domElement: HTMLElement;

    beforeEach(() => {
      domElement = makeDomElement(
        `<button
          ${insights('clickedRecordIDsAfterSearch', {
            ids: ['3'],
            eventName: 'Add to Cart',
          })}> Add to Cart </button>`
      );
    });

    it('should extract the method name', () => {
      expect(readDataAttributes(domElement).method).toEqual(
        'clickedRecordIDsAfterSearch'
      );
    });

    it('should extract the payload and parse it as a json object', () => {
      expect(readDataAttributes(domElement).payload).toEqual({
        ids: ['3'],
        eventName: 'Add to Cart',
      });
    });
  });
});
