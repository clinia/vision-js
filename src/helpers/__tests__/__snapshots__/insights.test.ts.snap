// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insights default behaviour 1`] = `"data-insights-method=\\"clickedRecordIDsAfterSearch\\" data-insights-payload=\\"eyJpZHMiOlsiMyJdLCJldmVudE5hbWUiOiJBZGQgdG8gQ2FydCJ9\\""`;

exports[`readDataAttributes on handwritten data-insights-* attributes should reject invalid payload 1`] = `"The insights helper was unable to parse \`data-insights-payload\`."`;

exports[`writeDataAttributes should output a string containing data-insights-* attributes 1`] = `"data-insights-method=\\"clickedRecordIDsAfterSearch\\" data-insights-payload=\\"eyJpZHMiOlsiMyJdLCJldmVudE5hbWUiOiJBZGQgdG8gQ2FydCJ9\\""`;

exports[`writeDataAttributes should reject non JSON serializable payloads 1`] = `"Could not JSON serialize the payload object."`;

exports[`writeDataAttributes should reject non object payloads 1`] = `"The insights helper expects the payload to be an object."`;

exports[`writeDataAttributes should reject undefined payloads 1`] = `"The insights helper expects the payload to be an object."`;
